[WIP] Roadmap item: platform categories

levitte opened the pull request:
Our roadmap says this:

> Platform Strategy
> 1.  Classify all the platforms known by Configure according to the  [platform policy](https://www.openssl.org/policies/platformpolicy.html) (Timescale: Next feature release)
> 2.  Publish the build and test status for each platform (Timescale: Next feature release)

This PR attempts to address item 1.

levitte committed:
Propose a numbering scheme for the config targets categories

This is according to our current roadmap, which says among others:

Platform Strategy

    1. Classify all the platforms known by Configure according to the
       [1]platform policy (Timescale: Next feature release)

References:

  1. https://www.openssl.org/policies/platformpolicy.html

levitte committed:
Boldly move the main config targets to category "unknown" for further eval

levitte committed:
Move the primary linux targets to 10-linux.conf

This is according to platform policy, but note that we're not moving
ALL Linux config targets to this file, only those we know that we use
and work on on a daily basis.

levitte committed:
Move fallback / minimal config targets to their own file

These targets fall outside our platform strategy, as they aren't tied
to any specific platform, and are rather to be seen as a fallback or
the bare minimum that should work on any platform with a Unix-like
user interface

levitte commented:
The main activity here should be to look through `Configurations/70-main.conf` and classify the config targets there into:

- secondary (if someone on @openssl/committers is willing to actively support that set of configs)
- community (if it's known that someone in the community uses this set of configs and is likely to help us fix issues with it)
- deprecated (if it's known that this set of configs isn't really used anywhere, i.e. the platform is unused and unsupported / unavailable, ...)
- unknown (if we can't know for the moment)

The goal is to reduce the config targets in the "unknown" category to a minimum, i.e. we should make an effort to place each config target in another category.
At the same time, I personally feel that it should be fairly hard to deprecate a config target.  If you do indicate a deprecation, expect to have to answer to "Really?  Really really???" type of questions.  We *are* known for trying to maintain config targets for old / little known platforms, I think that's a valuable attitude to maintain.

t-j-h commented:
Pull the WinCE platforms out separately.
They are not currently "secondary".

levitte commented:
Okie

richsalz commented on Configurations/70-windows-msvc-wince.conf:
```diff
@@ -0,0 +1,121 @@
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+# Helper functions for the Windows configs
```
Should this be Windows-CE configs?

levitte replied:
Ah, good point.

richsalz commented on Configurations/README:
```diff
@@ -15,6 +15,30 @@ configuration in diverse ways:
+Config files (*.conf) are named {nn}-{name}.conf, where {nn} is a two
+digit number and {name} is an indicator of what kind of configuration
+can be expected in the file.  The convention for {name} may be refined
+for some number groups.
```
Perhaps add: "The numbers have no significance other than grouping. All config names across all files must be unique"

levitte replied:
Okie (or something along those lines at least), good idea

levitte replied:
Ah, I don't agree that names must be unique across all files.  For example, we have 10-linux.conf, which is the small set of linux configs we view as primary, but there is a number of them that we cannot claim, and I fully expect those to end up in, say, 50-linux.conf.  Why complicate matters further?

richsalz replied:
Not the filenames, the configuration target names.

levitte replied:
Ahhhh, yeahok, I'll put the wording in

richsalz commented on Configurations/README:
```diff
@@ -15,6 +15,30 @@ configuration in diverse ways:
+(*) See https://www.openssl.org/policies/platformpolicy.html
+(**) There may be some configs that team members use personally and
+     wish to share.  This also leaves space for those who want to
+     maintain their own configs inside their OpenSSL source checkout
```
Should we define a number/name that we will *NEVER* use, so that downstream folks don't have to work.  For example, reserve 88-xxx for others to use.

levitte replied:
I think I'd rather add a section.  Perhaps steal 70-79 from "Unknown", say?

dot-asm commented on Configurations/20-windows-posix.conf:
```diff
@@ -0,0 +1,111 @@
+
+(
+#### MinGW
+    "mingw" => {
```
mingw is *not* "windows posix" thing, it straight windows, just non-ms compiler.

levitte replied:
With the mingw shell, it does provide a POSIXly environment, no?  But feel free to suggest a better file name...

dot-asm replied:
There is no "mingw shell". In most cases it's MSYS2 shell with mingw cross-compiler on $PATH. [In the context note that MSYS2 is windows posix" thing, and it *is* different from mingw.] As for cases other than most, it can as well be Linux shell with mingw cross-compiler on $PATH. That's how it's done on travis.

As for name suggestion, how about 20-windows,conf with ms and mingw targets.

levitte replied:
So you're suggesting on `20-windows.conf` containing all the (non-CE) VC- config targets and all the mingw config targets.  Do I get that correctly?  What do you want done with cygwin?

dot-asm commented:
> deprecated

All SCO/unixware targets. One can probably argue that it's time to classify Tru64 and IRIX targets as deprecated too [on ground of lack of vendor support]... Though there is ambiguity. "deprecated" are ranked "lower" than "unknown", but Tru64 and IRIX are actually *known* to work. MPE on the other hand is unknown, and so is nextstep and BS2000.

> secondary

I effectively "actively supports" all platforms with assembly packs. "Effectively" means that I don't actually test for example BSD, but I see no reason to believe that assembly code wouldn't work there. And problem reports will be addressed.

> community

There were some activity signs in qnx field. Normally vos would go there too, but there were no signs of activity since 1.0.2...

dot-asm commented:
vxworks... basically same as wince...

levitte commented:
> Though there is ambiguity. "deprecated" are ranked "lower" than "unknown", but Tru64 and IRIX are actually *known* to work.

I'm not sure there's any value in trying to see it as a rank, but rather as an indication what we might do with them in the long run.  I also think that vendor support is completely irrelevant if there's community support of some sort...  after all, if they are *known* to work, it should mean that someone in the community has taken the time to test them a little now and then.

levitte commented:
> vxworks... basically same as wince...

So, errr, unknown?  That's where WinCE is at this moment...

dot-asm commented:
> Tru64 and IRIX are actually known to work.

I'm speaking for myself here.

levitte commented:
> "Effectively" means that I don't actually test for example BSD, but I see no reason to believe that assembly code wouldn't work there. And problem reports will be addressed.

That doesn't make BSD a secondary platform.  For a platform to be secondary, it means that someone in project team has access to that platform and is actively building and testing OpenSSL on it.  (what "actively" means is self defined, but I'd say that at a minimum try to build and run tests with some regularity, and fix issues when they show up)

levitte commented:
> > Tru64 and IRIX are actually known to work.

> I'm speaking for myself here.

So does that mean that you have personally tested it them?  Have access to them?

kaduk commented:
> That doesn't make BSD a secondary platform. For a platform to be secondary, it means that someone in project team has access to that platform and is actively building and testing OpenSSL on it.

I have access to FreeBSD and can move some of my normal workload onto it.

levitte committed:
Move SCO/Unixware targets to their own file

Category: deprecated according to platform policy

levitte commented:
@kaduk, note that FreeBSD is very much bundled together with the other BSDs.  [here](https://github.com/openssl/openssl/blob/master/config#L690-L707) is the section in [config](https://github.com/openssl/openssl/blob/master/config) that auto-selects the BSD-related config targets.  So looking at those, I think you have the choice of taking them all on in bulk, or select the processors you have on hand.
(note that there exists a config target `bsdi-elf-gcc` that's kind of a little beast of its own that's still actively tested by one user I know of...  don't bother with that one)

mspncp commented:
Have you already considered posting an inquiry to the openssl-user mailing list asking for a sign of  life for the platforms which are in an unknown state? You could ask for volunteers that are willing to keep their favorite  platforms alive.

dot-asm commented:
>>> Tru64 and IRIX are actually known to work.

>> I'm speaking for myself here.

> So does that mean that you have personally tested it them? Have access to them?

Yes. But there is uncertainty in whether or not it would be case tomorrow. Obviously not *literally* tomorrow, but nevertheless... Though it should be noted this uncertainty applies even to HP-UX...

levitte committed:
Move QNX targets to their own file

Category: community according to platform policy
(based on "There were some activity signs")

levitte commented:
> Have you already considered posting an inquiry to the openssl-user mailing list asking for a sign of life for the platforms which are in an unknown state? You could ask for volunteers that are willing to keep their favorite platforms alive.

Yes, we should.  What we're doing here is a first triage...

levitte commented:
One part of the triage involves finding what platforms We The Team want to support (i.e. become secondary).

kaduk commented:
@levitte If the BSDs get the same configure stanza, then testing on one of them implies that the stanza is tested and that should suffice, no?  Alternately, we could split things out to reflect what is actually tested...

levitte commented:
> Yes. But there is uncertainty in whether or not it would be case tomorrow. Obviously not literally tomorrow, but nevertheless... Though it should be noted this uncertainty applies even to HP-UX...

I get you about uncertainty, and I assume you're saying that you can never know half a year from now ("now" being any moment in time, basically).

So tell you what...  if *I* was in your shoes, I would at least consider that community support (because we're all part of a development community, regardless of if we're on the OMC, if we have commit access or whatnot).  and also, if *I* thought I could actively support a platform for a year from now, I would consider having it categorised as "secondary", for the time being.  But that's entirely up to you; this is a voluntary categorisation, after all, based on what each of us is willing to put up with.

levitte commented:
> @levitte If the BSDs get the same configure stanza, then testing on one of them implies that the stanza is tested and that should suffice, no? Alternately, we could split things out to reflect what is actually tested...

Well yes, roughly speaking you are correct.  I think the biggest reason to have them divided along the processor axis is assembler, and those do differ, obviously.  From that perspective, it does make sense for you to be selective

mspncp commented:
You could create  a wiki list of current „platform maintainers“ and their GitHub handles, which is easy to update when things change.

levitte commented:
> You could create a wiki list of current „platform maintainers“ and their GitHub handles, which is easy to update when things change.

That could be an idea...  but it does depend what we mean with "community supported"...  after all, a config target could be proposed at some point in time, and even though the person is all chipper at the time and seems to want to support the platform for a while, we cannot know what things will look like half a year later.  However, even if *that* person has gone awol, *someone else* may very well take up the gauntlet some time and fix what needs fixing if there is any...  or we simply won't hear a peep for a few years.  So "community supported" is realistically a very dynamic thing.

That, or perhaps we need to look at "community supported" more critically, and really only place platforms there that do get long term consistent support by, errrr, someone, and thereby have a whole bunch in "unknown", 'cause even if there is activity around them, it may be fairly random, inconsistent, whatnot.

It's all in one's point of view, and as of this moment, I'm undecided which way to go.  Opinions welcome.

levitte commented:
For the "secondary" category, it's much much easier, We The Project Team are expected to be present and active over longer term, and not go AWOL.  The main supporters of each set of platforms are noted in the config file itself.

mspncp commented:
> That, or perhaps we need to look at "community supported" more critically, and really only place platforms there that do get long term consistent support by, errrr, someone, and thereby have a whole bunch in "unknown", 'cause even if there is activity around them, it may be fairly random, inconsistent, whatnot.

> For the "secondary" category, it's much much easier, We The Project Team are expected to be present and active over longer term, and not go AWOL. The main supporters of each set of platforms are noted in the config file itself.

A compromise could be to rename the "community supported" category to "community" and join it with the "unknown" category. Then the script files don't have to be renamed every time the state changes.

For the officially supported  and the "secondary" targets, the supporter is in listed in the config file, whereas the "community" platform supporters are listed only in the wiki, which acts as a self help bulletin board.

mspncp commented:
@t-j-h mentioned companies in https://github.com/openssl/openssl/pull/5035#issuecomment-356233430, that gave up on upstreaming their platform specific patches because their requests remained unheared for a long time:

> The companies that have done this haven't bothered with trying to upstreaming their patches after they were rejected when they tried long ago (multiple times). The compilation passes (and pretty much always has) with very minor changes. 

This was probably in the pre-Heartbleed era and one could try to contact these companies and reactivate them, if they are still interested in contributing.

levitte committed:
Move darwin-x86_64 to its own file

Category: secondary according to platform policy

dot-asm commented:
As already said, one can argue that Tru64 and IRIX should be classified as deprecated. Let me rephrase, *I* argue that Tru64 and IRIX should be classified as deprecated. If somebody is willing to step up, they are free to do so.

t-j-h commented:
That categories are defined in our policy. So we need to use them as defined and see how it turns out. We had this debate at length. 

There isn't a category of "works" or "doesn't work" deliberately. This is about what a user can expect in terms of dealing with issues.

Primary and secondary are a commitment from the project team ... that we actively "support" the platform in the sense that we use it regularly so it should be safe to depend on as a platform. 

Most current platforms belong in the "unknown" category with a few in "deprecated" where we think they are not widely used and we think we may indeed remove them in the future.

levitte commented:
> As already said, one can argue that Tru64 and IRIX should be classified as deprecated. Let me rephrase, *I* argue that Tru64 and IRIX should be classified as deprecated. If somebody is willing to step up, they are free to do so.

Ah, ok, I hear you.

levitte committed:
Move BSD-x86 and BSD-x86_64 to their own file

Category: secondary according to platform policy

levitte committed:
Move Irix and Tru64 to their own files

Category: deprecated according to platform policy

levitte commented:
@dot-asm, does the Windows / Mingw / Cygwin organisation look better to you now?

levitte committed:
Divide remaining config targets into files divided by operating system

Category: unknown according to platform policy, as we have no idea
what the status is for all these config targets as of now.

levitte committed:
Move VMS targets to their own file

Category: secondary according to platform policy

levitte committed:
Move Cygwin config targets to their own file

Category: secondary according to platform policy

levitte committed:
Move Windows CE config targets to their own file

Category: unknown according to platform policy

levitte committed:
Move Windows (MSVC) and Mingw config targets to their own file

Category: secondary according to platform policy

levitte commented:
I've split of the rest of what was 10-main.conf into a number of files, all in group 80 ("unknown").  I'd say that pretty much concludes this PR.

I've just one last question...  Cygwin, I've currently classified it as "secondary", but am thinking it should move to "community".  The reason is that no one in the team really gives it much day to day care.  I know that both @mattcaswell and I test it before releasing (and find that it builds and tests without issues), but other than that, at least *I* don't even look, and I wouldn't be surprised of the larger community would find issues earlier than We The Project Team do.
Thoughts?

t-j-h commented:
Cygwin is *not* a secondary platform *unless* a team member steps forward to "own" it. The platform policy has mingw as secondary which is specifically not Cygwin. Please put Cygwin into unknown.

richsalz commented:
@dot-asm has worked with them to keep things current (Corinna at RedHat I think).  I also helped do some reviews and merges.  I would leave it as secondary.

levitte commented:
Unknown is clearly wrong, as several of us keep an eye on it.  It is clearly known.

t-j-h commented:
Someone has to step forward to either make it secondary or someone in the community - or it is "unknown". It is not currently a secondary platform.

dot-asm commented:
> does the Windows / Mingw / Cygwin organisation look better to you now?

Mingw, yes. As for Cygwin, one can wonder if you would be appropriate to call the file e.g. posix_on_win. Rationale is that there is remote possibility that msys2 would be added. On the other hand one can as well argue that it would be as appropriate to have it separate, as it's not unlike if it will classified differently and would end up in separate file... I'm leaning toward leave it, cygwin, as it is now, in this request.

dot-asm commented:
Once again, I effectively "actively supports" all platforms with assembly packs. All right, I don't check BSD, but I regularly check linux and android targets. As well as AIX. In other words these are *not* "unknown", but "secondary". Occasionally I check even HP-UX, but one can probably say not often enough to classify it higher than "community".

dot-asm commented:
Note about ios targets. There is no option to run complete test suite, but each ARM assembly modification is exercised by manually executing relevant tests.

t-j-h commented:
This isn't about "works" - this is about "supported". And someone has to volunteer to *support* a platform.
And occasionally checking when you change ASM code wouldn't really match my definition of "supports".

dot-asm commented:
As for targets like linux-arm and linux-aarch64. It's not really inappropriate to consider them popular/important enough to deserve a bit special treatment. As pointed out in various private conversations, it's perfectly possible to set up cross-compiled *and* cross-execution environment at `run` host, so that they are exercised automatically on daily basis. Here is question, how would it affect classification?

t-j-h commented:
If you are willing to actively support those platforms (and your description of automated daily builds seems like that) then they would become secondary platforms under the platform policy.

richsalz commented:
Let's also keep in mind that almost everything is best effort, and that we wrote the policy having had absolutely no experience in the project with this.  So I think we should focus on making it better, not necessarily correct (let alone perfect).

rsbeckerca commented:
Greetings, I have been asked by the ITUGLIB team to reach out to the OpenSSL group and offer Community-level support for the OpenSSL port on the NonStop NSE, NSX, and NSV platforms. We are more than willing to contribute our port to the main thread. Our port branch is off e5bba24cd8bb3e5127a578b85c6edf013a38ea6d.

paulidale commented:
We've got a number of platforms we do nightly builds for.  Solaris failed about a month back and I've only just submitted a PR to fix it.  Is that timely enough to qualify as actively supported?

levitte commented:
@paulidale, that would be nice :-)

levitte commented:
@rsbeckerca, cool!  I suggest a PR that includes a file `Configurations/50-nonstop.conf` with the appropriate settings.  If you need to discuss it first, or need help (for example to write a proper config file), I suggest starting with a [github issue](https://github.com/openssl/openssl/issues).

paulidale commented:
I'll make a note to create a list of them all and figure out where they fit with respect to the configuration scripts.

paulidale commented:
We do nightly builds for these machines/architectures:

- `AIX PPC32`
- `AIX PPC64`
- `HPUX IA64`
- `HPUX IA64`
- `Linux PPC` Little Endian
- `Linux s390` 32 bit
- `Linux s390` 64 bit
- `Linux Sparc`
- `Oracle Linux 5 x86-64`
- `Oracle Linux 5 x86`
- `Solaris Sparc v8p`
- `Solaris Sparc v9`
- `Solaris x86-64`
- `Solaris x86`
- Various Windows builds

levitte commented:
You know, I'm starting to think that this number group thing wasn't the best idea.  With @paulidale's comment, we would for example end up with `10-linux.conf` (contains the generics and linux on x86 and x86_64), `20-linux.conf` (containing a *subset* of our Linux on PPC config targets, along with Linux on s390 and Linux on Sparc), and `80-linux.conf` (*the rest of* Linux on PPC and Linux on other odd platforms).

Frankly, that seems more confusing than helpful.  Sure, we can invent some more helpful names, but with the division of Linux on PPC, we'd get a `20-linux-ppc64le.conf` and then what, `80-linux-remaining.conf`?

So I'm wondering if classification shouldn't be done with an attribute in the configs instead:

-   For primary:

    ```
            category => 'primary'
    ```

-   For secondary:

    ```
            category => 'secondary',
            supporters => [ 'someone@somewhere.com', 'someoneelse@whatever.org', ... ],
    ```

-   For community:

    ```
            category => 'community',
            supporters => [ 'someone@somewhere.com', 'someoneelse@whatever.org', ... ],
    ```

-   For "unknown":

    ```
            category => 'unknown',
    ```

-   For legacy:

    ```
            category => 'legacy',
    ```

thoughts?

levitte commented:
It would be perfectly possible to have `Configure` list those data with specific options, just like we already have `LIST` to list the current config targets

kaduk commented:
I think that having level-of-support split out into separate files is probably a helpful signal to readers.  But that's not to say that we need to keep number groups or limit ourselves to just ```foo-linux.conf``` -- why not ```linux-tier1.conf```, ```linux-community.conf```, etc?

levitte commented:
Well, I'm thinking of `linux-ppc`, `linux-ppc64` and `linux-ppc64le` that would be split up between two files if I assign just the last one into a tier2 and keep the rest as "unknown"...  That's not particularly helpful for someone that wants to look over our configs for Linux for PPC in one go.  It's a maintenance problem waiting to happen.

rsbeckerca commented:
Personally, I've been a long-time fan of a reverse DNS-analogue for even pseudo hierarchies like this, so something like `linux.ppc`, `linux.ppc.64`, `linux.ppc.64.le`, `nonstop`, `nonstop.nsx`, `nonstop.nse`, `mac`, `mac.osx`, etc.

levitte commented:
Did that have anything to do with file naming, i.e. a suggestion that there should be files like `linux.ppc.conf`, `linux.ppc.64.le.conf`, `nonstop.conf`, ... ?  Also, that doesn't answer how to classify them according to our platform policy...

rsbeckerca commented:
That was my suggestion, yes, instead of the numbering approach, to use linux.ppc.conf, but then if we have a community definition separate, having something like `linux.community...conf`. I was being speculative and incomplete. apologies.

rsbeckerca commented:
I was wondering whether there was an update to the very long link line situation that causes NonStop's linker to become upset.

richsalz commented:
Probably better to open a separate issue and put as much information in there as you can.  The link command is separate from the config filenames.  And nobody has looked at it.  Someone with a NonStop system will have to work on it.

rsbeckerca commented:
Split off to issue #5335 as requested.

levitte commented:
@richsalz, FYI: #5087

levitte commented:
Closing this one for the benefit of #5381
